### ISSUE TITLE
Deployment Docs

### ISSUE THREAD

**Issue by philippjfr** [holoviz/panel#537]
We should have detailed instructions on deploying Panel dashboards/apps on different platforms and embedded in different frameworks. If anyone can suggest any more topics, platforms and frameworks please chime in and I'll add them to the list.

## Topics

- [ ] Authentication

## Platforms

- [ ] AWS
- [ ] Google Cloud
- [x] Heroku
- [ ] Digital Ocean
- [ ] Azure
- [x] Binder (or just myBinder?)
- [ ] Anaconda Enterprise

## Frameworks

- [ ] Flask
- [ ] Django

**Comment by sametz** [comment]
In the FAQ tool comparison chart, for "Servable from public site", in the "Panel" column it says ("Yes; mybinder Heroku, etc."). I've been looking for more info on the mybinder part. If there's a way to serve a Panel app through a GitHub repo, similar to using  binder with a Jupyter notebook, or using nbinteract, instructions on how to do that would be VERY useful.

**Comment by jbednar** [comment]
@sametz , sure; see https://github.com/pyviz-demos/glaciers (and other examples at https://github.com/pyviz-demos), click the button ![image](https://user-images.githubusercontent.com/1695496/61494572-59851700-a96b-11e9-849f-7a972da2b059.png)
and then look through the repo to see how it works.

**Comment by sametz** [comment]
@jbednar Thanks! This could be a gamechanger for me.

**Comment by jbednar** [comment]
I think we should explicitly add a couple of baselines before getting into any of the ones above:

- [ ] Local usage (open a command prompt and run `panel serve` or `jupyter notebook`)
- [ ] Remote usage (open an ssh session and run `panel serve`, or set up an ssh tunnel, run `jupyter notebook`, and connect to the tunnelled port)

We can say that these two approaches should work on any systems that you can get ssh access to, and that for these and for any of the others you will need to think about authentication. Then we can talk about how the other options compare to these baselines, i.e. how they can spin up containers, allocate cloud resources automatically, etc.  

Not sure if we should mention running in docker manually, which would be in between these two approaches and the rest of them; I assume that would not be a common thing to want to do.

**Comment by jbednar** [comment]
Oh, and here I'm assuming we should be discussing both deployments and live, running sessions; some of the same considerations apply in both cases, while others differ.

**Comment by philippjfr** [comment]
@ericml Has put together a minimal panel app (https://github.com/ericmjl/minimal-panel-app) and deployed it to Heroku (http://minimal-panel-app.herokuapp.com/minimal-panel). Should make a great template for writing docs.

**Comment by sametz** [comment]
I finally succeeded with a GitHub deploy, but it took this newbie many, many hours struggling with fiddly bits (but I can see that someone with experience with Bokeh and deployment would find this trivial). So, IMO adding "GitHub" to the list of Platforms in the OP would be very helpful.

For reference, here's my test app: https://github.com/sametz/panel-test

I'm impressed how well this works on a smart phone. This could be very useful for my students.

**Comment by jbednar** [comment]
For AWS, can consider capturing the setup using https://aws.amazon.com/cloudformation .

**Comment by marfel** [comment]
So... how did your PyConDE Sprint turn out? I would be particularly interested in the "authentication" topic from the first post: We have a case here where Panel would be the perfect tool to design a front-end for a data science project, but we must be able to deploy it on our own server while allowing only authenticated users access to it. Not being a web guy myself, I would prefer not having to set up an entire Django server or the like, just to provide authentication (if that is even possible). That is, unless there is step-by-step documentation for this use case. Having control over user access would increase the range of possible Panel webapps tremendously for us!

**Comment by hungyiwu** [comment]
tl;dr please add more documentation to the server deployment, especially AWS!

I use python for a few years and started to build an interactive dashboard a week ago for remote collaboration. Tried Dash but it assumes a lot of HTML and CSS knowledge in order to layout properly. Discovered Panel from a SciPy conference talk on YouTube and loved the simple layout (pn.Column, pn.Row). Struggled for a week to deploy it to AWS Elastic Beanstalk. I know Heroku works, but my data sits on AWS S3. Will leave it for now, checking back sometime in the future.

**Comment by konradsitarz** [comment]
Hey, I try to deploy the Panel dashboard on Azure with AAD authorization.
I set up all the environment variables as in the img below
![image](https://user-images.githubusercontent.com/51296911/97331735-fd9d8380-1879-11eb-8a2c-4dbc2cf0b4ad.png)

my starting command looks like this 
`panel serve --allow-websocket-origin=appname.azurewebsites.net --port 8888 mydashboard.py`

Dashboard starts but no auth is set up. In the logs I see:
`User authentication hooks NOT provided (default user enabled)`

Could you please give some advise what can be wrong here?

**Comment by WesleyTheGeolien** [comment]
Hi guys,

I have made a successful Google Cloud deployment and would be happy to write the docs / steps I took for that if that is of any use to anyone?

**Comment by MarcSkovMadsen** [comment]
Hi  @WesleyTheGeolien 

Deployment docs would be so nice to have.

**Comment by WesleyTheGeolien** [comment]
@MarcSkovMadsen Just taken a quick look at the `panel/doc(/user_guide/index.rst)` and couldn't immediately see where the server deployment documentation is being pulled from and chance of pointing me in the right direction?

**Comment by xavArtley** [comment]
Most documentation in generated through notebooks
for example https://panel.holoviz.org/user_guide/Deploy_and_Export.html is contained in https://github.com/holoviz/panel/blob/master/examples/user_guide/Deploy_and_Export.ipynb

**Comment by WesleyTheGeolien** [comment]
Awesome thanks @xavArtley that is exactly what I was looking for 👍

**Comment by t-houssian** [comment]
@philippjfr Another helpful framework would be FastAPI (https://fastapi.tiangolo.com/) I have been trying to integrate it and can't quite seem to get it.

**Comment by manuvazquez** [comment]
> I finally succeeded with a GitHub deploy, but it took this newbie many, many hours struggling with fiddly bits (but I can see that someone with experience with Bokeh and deployment would find this trivial). So, IMO adding "GitHub" to the list of Platforms in the OP would be very helpful.
> 
> For reference, here's my test app: https://github.com/sametz/panel-test
> 
> I'm impressed how well this works on a smart phone. This could be very useful for my students.

This is pretty old and, seemingly, this is _checked_ as working, but I'm a bit lost.  Could you please elaborate  a little more? It's not as straightforward as stated in the [documentation](https://panel.holoviz.org/user_guide/Server_Deployment.html#mybinder), right? Also, when I click on the "launch binder" badge in your repo, sometimes it works nicely but others I get `[Errno 99] Cannot assign requested address` (in that case, it seems refreshing the webpage from the browser is enough for the app to load, though).

**Comment by jbednar** [comment]
There's a doc for deploying on Google Cloud at: https://towardsdatascience.com/deploy-a-python-visualization-panel-app-to-google-cloud-cafe558fe787

It would be great to curate these better for users somehow, but it's a bit much for us to keep track of all these options!

**Comment by sametz** [comment]
> This is pretty old and, seemingly, this is _checked_ as working, but I'm a bit lost. Could you please elaborate a little more? It's not as straightforward as stated in the [documentation](https://panel.holoviz.org/user_guide/Server_Deployment.html#mybinder), right? Also, when I click on the "launch binder" badge in your repo, sometimes it works nicely but others I get `[Errno 99] Cannot assign requested address` (in that case, it seems refreshing the webpage from the browser is enough for the app to load, though).

It's been a while since I looked at this demo. I also was getting the same error; even on the most recent branch and rebuilding the environment. My first guess is one of the dependencies that's not pinned to a specific version now needs to be for compatibility. I've been meaning to get back to my work on NMR apps, so this may be a motivating example to resume work.

**Comment by manuvazquez** [comment]
> There's a new doc for deploying on Google Cloud at: https://towardsdatascience.com/deploy-a-python-visualization-panel-app-to-google-cloud-cafe558fe787
> 
> It would be great to curate these better for users somehow, but it's a bit much for us to keep track of all these options!

Google Cloud huh...interesting :smiley:  thanks for the pointer!!

**Comment by jbednar** [comment]
I do think it makes sense to document how to use Panel within Flask. Flask doesn't do anything for Panel that I know of, but I can definitely imagine someone with a Flask app wanting to incorporate some Panel components, which seems reasonable to me!

**Comment by mhangaard** [comment]
A code example of how to set up routing correctly in Panel, so the OAuth flow works would be greatly appreciated :) I had to read through the issues to find out why Panel redirects to relative /login by default, while setting up OAuth with Azure.

**Comment by atisor73** [comment]
I recently tested Panel on an AWS EC2 instance, and listed my steps here: https://github.com/atisor73/panel-aws-tutorial. The commands are similar to deploying on Azure, and setting up the AWS instance was similar to existing guides on deploying Streamlit apps. 

I would be happy to submit a PR adding to the documentation. I'm not sure what else to include, if anyone had any thoughts let me know - I have not Dockerized the app or added a proxy server (Apache/Nginx), and haven't connected to any databases, so it's a very bare-bones example of just configuring the ports.

**Comment by philippjfr** [comment]
@atisor73 sorry I totally missed your reply here. I'd love to see a PR, any improvement to the AWS deployment docs is a welcome one.

I'll close this issue for now, we now cover a variety of options and future additions should be handled via new individual issues.
